src/platform/posix/tcp_ip_channel.c

```diff
@@ -411,41 +457,56 @@ static void *_TcpIpChannel_worker_thread(void *untyped_self) {
   }
 
   while (!self->terminate) {
-    switch (_TcpIpChannel_get_state(self)) {
-    case NETWORK_CHANNEL_STATE_OPEN: {
-      /* try to connect */
-      if (self->server) {
-        _TcpIpChannel_try_connect_server(untyped_self);
-      } else {
-        _TcpIpChannel_try_connect_client(untyped_self);
-      }
-    } break;
-
-    case NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS: {
-      _env->platform->wait_for(_env->platform, self->super.expected_try_connect_duration);
-    } break;
-
-    case NETWORK_CHANNEL_STATE_LOST_CONNECTION:
-    case NETWORK_CHANNEL_STATE_CONNECTION_FAILED: {
-      _env->platform->wait_for(_env->platform, self->super.expected_try_connect_duration);
-      _TcpIpChannel_reset_socket(self);
-      _TcpIpChannel_update_state(self, NETWORK_CHANNEL_STATE_OPEN);
-    } break;
-
-    case NETWORK_CHANNEL_STATE_CONNECTED: {
-      ret = _TcpIpChannel_receive(untyped_self, &self->output);
-      if (ret == LF_OK) {
-        validate(self->receive_callback);
-        self->receive_callback(self->federated_connection, &self->output);
-      } else {
-        LF_ERR(NET, "TcpIpChannel: Error receiving message %d", ret);
+    pthread_mutex_lock(&self->state_mutex);
```

Comment:
This means that we cannot send concurrently with receiving? It seems a bit too constraining?